### ISSUE TITLE
Cost function of message retrieval

### ISSUE THREAD

**Issue by s-tikhomirov** [waku-org/research#35]
What should determine the cost of retrieving a message?

Possible approaches:

- constant per message
- constant per byte
- dependent on message age (presumably, newer messages are cheaper)
- dependent on other message parameters (topic, metadata, etc)

**Comment by fryorcraken** [comment]
With the new store req-res that returns message ids, then it becomes trivial to have a cost per message.
This assumes that client does not pay for message ids, but pays for message content:

```mermaid
sequenceDiagram
	client-->>server: request msgs ids for period t1 to t2
	server-->>client: returns 10 msg ids
	client->>client: check seen msgs
    client->>client: 3 msgs are missing
	client->>blockchain: pays server for 3 msgs
	client->>server: request 3 messages, attaches txid
	server->>blockchain: checks payment
	server-->>client: deliver messages
```

PS: sorry about the label deletion, this PR will remove this kind of automtaed action: https://github.com/waku-org/pm/pull/103

**Comment by s-tikhomirov** [comment]
> client does not pay for message ids, but pays for message content

The scheme makes a ton of sense!

Do you think we should aim for it in the PoC, or in later protocol iterations? As of the current PoC outline, the client skips the request-ids stage and instead pays a fixed price per hour of history, and gets all relevant messages in response, no matter if the client already has them or not. This is kinda wasteful (client gets messages it already has) but conceptually simpler (just one round-trip).

**Comment by fryorcraken** [comment]
> * to compare responses by different store nodes and choosing the one with the most up-to-date response

Absolutely, I think I wrote somewhere how Alice could then do a sync req to Bob and Carole.
Compare what messages they have, query/pays Bob because he's cheapest to retrieve the bulk of messages, and then query/pays Carole of a unique message she has, but she is more expensive.

It **may** also make the market more interesting:

- Bob is cheaper because he only store 12 hours of messages.
- Carole is more expensive because she stores a week of message.

Alice will always use Bob for the most recent messages, but has to use Carole if she was offline for more than 12 hours.

What is interesting here, is that once the protocol is available, this is purely an engineering problem of window-shopping and cherry-picking.



> > client does not pay for message ids, but pays for message content
> 
> The scheme makes a ton of sense!
> 
> Do you think we should aim for it in the PoC, or in later protocol iterations? As of the current PoC outline, the client skips the request-ids stage and instead pays a fixed price per hour of history, and gets all relevant messages in response, no matter if the client already has them or not. This is kinda wasteful (client gets messages it already has) but conceptually simpler (just one round-trip).

I will let you and @jm-clius decide here.

My current opinion is that when store sync is available, then I assume the best practice to be:
(A)
- store sync
- store query missing messages using id.

instead of
(B)
- store query using using time frame or cursor and get full messages.

My expectations might be incorrect tho. My assumption is that (A) gives you a better bandwidth/latency ratio.

Note: when I write "store sync" I mean "store query to retrieve messages id". This may not be the correct/final name of the protocol.

**Comment by s-tikhomirov** [comment]
> My current opinion is that when store sync is available, then we can assume the best practice will become:
(A)
store sync
store query missing messages using id.

My tangential concern here is about the relationship between Store sync and incentivized Store requests. If both are available, why would a rational client use incentivized requests? It can "pretend" to be a syncing server and get the same data for free. I know there may be heuristics to distinguish such clients-pretending-to-be-servers, but I'm not sure we can do it with certainty.
